{"cells":[{"cell_type":"markdown","id":"f9d7fb46-e640-47fc-a7ad-7488ca5d2a6a","metadata":{"id":"f9d7fb46-e640-47fc-a7ad-7488ca5d2a6a"},"source":["1. Write a Python program to handle a ZeroDivisionError exception when dividing a number by zero."]},{"cell_type":"code","execution_count":null,"id":"f1969a88-be36-494b-af4d-4595ef3173a6","metadata":{"id":"f1969a88-be36-494b-af4d-4595ef3173a6","outputId":"37076c97-eab7-4fa2-e825-3c95dff7eb6a"},"outputs":[{"name":"stdout","output_type":"stream","text":["Error: Cannot divide by zero\n"]}],"source":["def divide_numbers(x, y):\n","    try:\n","        result = x / y\n","        print(\"Result of division:\", result)\n","    except ZeroDivisionError:\n","        print(\"Error: Cannot divide by zero\")\n","\n","# Example usage:\n","numerator = 10\n","denominator = 0\n","\n","divide_numbers(numerator, denominator)"]},{"cell_type":"markdown","id":"36fc508e-27d8-463c-8e83-12e8371faf87","metadata":{"id":"36fc508e-27d8-463c-8e83-12e8371faf87"},"source":["2. Write a Python program that prompts the user to input an integer and raises a ValueError exception if the input is not a valid integer."]},{"cell_type":"code","execution_count":null,"id":"2cefad92-abd1-4173-9ecb-bdfcdb09bdf2","metadata":{"id":"2cefad92-abd1-4173-9ecb-bdfcdb09bdf2","outputId":"59900c5e-f1bf-435c-9c21-3ee208c60104"},"outputs":[{"name":"stdin","output_type":"stream","text":["Please enter an integer:  gh\n"]},{"name":"stdout","output_type":"stream","text":["Error: Please enter a valid integer\n"]},{"name":"stdin","output_type":"stream","text":["Please enter an integer:  ty\n"]},{"name":"stdout","output_type":"stream","text":["Error: Please enter a valid integer\n"]},{"name":"stdin","output_type":"stream","text":["Please enter an integer:  1\n"]},{"name":"stdout","output_type":"stream","text":["You entered: 1\n"]}],"source":["def get_integer_input(prompt):\n","    while True:\n","        try:\n","            value = int(input(prompt))\n","            return value\n","        except ValueError:\n","            print(\"Error: Please enter a valid integer\")\n","\n","try:\n","    user_input = get_integer_input(\"Please enter an integer: \")\n","    print(\"You entered:\", user_input)\n","except ValueError as ve:\n","    print(ve)"]},{"cell_type":"markdown","id":"3d4bf68d-2c5e-4c06-bc5d-bff56c9a9617","metadata":{"id":"3d4bf68d-2c5e-4c06-bc5d-bff56c9a9617"},"source":["3. Write a Python program that opens a file and handles a FileNotFoundError exception if the file does not exist"]},{"cell_type":"code","execution_count":null,"id":"88430c09-a240-42a0-8e77-7739acaf5ae6","metadata":{"id":"88430c09-a240-42a0-8e77-7739acaf5ae6","outputId":"436f3cfd-3455-4513-a25d-17ca44067289"},"outputs":[{"name":"stdin","output_type":"stream","text":["Enter the name of the file to read:  iii\n"]},{"name":"stdout","output_type":"stream","text":["Error: File not found\n"]}],"source":["def read_file(file_name):\n","    try:\n","        with open(file_name, 'r') as file:\n","            content = file.read()\n","            print(\"File content:\\n\", content)\n","    except FileNotFoundError:\n","        print(\"Error: File not found\")\n","\n","# Example usage:\n","file_name = input(\"Enter the name of the file to read: \")\n","\n","read_file(file_name)"]},{"cell_type":"markdown","id":"186512ed-6d31-44e5-9173-68beed63f58e","metadata":{"id":"186512ed-6d31-44e5-9173-68beed63f58e"},"source":["4. Write a Python program that prompts the user to input two numbers and raises a TypeError exception if the inputs are not numerical."]},{"cell_type":"code","execution_count":null,"id":"dbf5997b-3bcd-48a6-9e4e-e5c6f033708b","metadata":{"id":"dbf5997b-3bcd-48a6-9e4e-e5c6f033708b"},"outputs":[],"source":["try:\n","    num1 = float(input(\"Enter the first number: \"))\n","    num2 = float(input(\"Enter the second number: \"))\n","\n","    result = num1 + num2\n","    print(\"Sum of the two numbers:\", result)\n","except ValueError:\n","    print(\"Error: Inputs are not numerical\")"]},{"cell_type":"markdown","id":"f4d2c88f-8018-4161-9006-570ae7fa87c5","metadata":{"id":"f4d2c88f-8018-4161-9006-570ae7fa87c5"},"source":["5. Write a Python program that opens a file and handles a PermissionError exception if there is a permission issue.\n"]},{"cell_type":"code","execution_count":null,"id":"6cd378e9-215f-4c9f-b212-2bcb4dbf190d","metadata":{"id":"6cd378e9-215f-4c9f-b212-2bcb4dbf190d"},"outputs":[],"source":["def read_file(file_name):\n","    try:\n","        with open(file_name, 'r') as file:\n","            content = file.read()\n","            print(\"File content:\\n\", content)\n","    except FileNotFoundError:\n","        print(\"Error: File not found\")\n","    except PermissionError:\n","        print(\"Error: Permission denied to access the file\")\n","\n","# Example usage:\n","file_name = input(\"Enter the name of the file to read: \")\n","\n","read_file(file_name)"]},{"cell_type":"markdown","id":"d55de498-ef68-4d95-8eac-bf38e0920c7a","metadata":{"id":"d55de498-ef68-4d95-8eac-bf38e0920c7a"},"source":["6. Write a Python program that executes an operation on a list and handles an IndexError exception if the index is out of range."]},{"cell_type":"code","execution_count":null,"id":"2b417981-9c87-4438-8cef-2ba8e130dd0c","metadata":{"id":"2b417981-9c87-4438-8cef-2ba8e130dd0c","outputId":"f7149371-89ce-4b89-ce6b-5d8fd5309209"},"outputs":[{"name":"stdin","output_type":"stream","text":["Enter the index to retrieve from the list:  7\n"]},{"name":"stdout","output_type":"stream","text":["Error: Index is out of range\n"]}],"source":["def perform_operation_on_list(input_list, index):\n","    try:\n","        result = input_list[index]\n","        print(\"Element at index\", index, \":\", result)\n","    except IndexError:\n","        print(\"Error: Index is out of range\")\n","\n","# Example usage:\n","my_list = [1, 2, 3, 4, 5]\n","\n","try:\n","    index = int(input(\"Enter the index to retrieve from the list: \"))\n","    perform_operation_on_list(my_list, index)\n","except ValueError:\n","    print(\"Error: Please enter a valid integer index\")"]},{"cell_type":"markdown","id":"88e42520-25c4-4e87-950e-02a97015e5af","metadata":{"id":"88e42520-25c4-4e87-950e-02a97015e5af"},"source":["7. Write a Python program that prompts the user to input a number and handles a KeyboardInterrupt exception if the user cancels the input."]},{"cell_type":"code","execution_count":null,"id":"23b7af27-7e11-4974-8d58-a4d0fdf22748","metadata":{"id":"23b7af27-7e11-4974-8d58-a4d0fdf22748","outputId":"f410221a-f54c-44c8-cdf7-c4f38337fff6"},"outputs":[{"name":"stdin","output_type":"stream","text":["Please enter a number:  \n"]},{"name":"stdout","output_type":"stream","text":["Error: Please enter a valid number\n"]}],"source":["try:\n","    user_input = input(\"Please enter a number: \")\n","    number = float(user_input)\n","    print(\"You entered:\", number)\n","except ValueError:\n","    print(\"Error: Please enter a valid number\")\n","except KeyboardInterrupt:\n","    print(\"\\nOperation canceled by user\")"]},{"cell_type":"markdown","id":"1373a5ba-db05-4bbe-b9d8-2f45ec774e50","metadata":{"id":"1373a5ba-db05-4bbe-b9d8-2f45ec774e50"},"source":["8. Write a Python program that executes division and handles an ArithmeticError exception if there is an arithmetic error."]},{"cell_type":"code","execution_count":null,"id":"a193858b-c90b-4c40-99bf-9e71404193ac","metadata":{"id":"a193858b-c90b-4c40-99bf-9e71404193ac","outputId":"10cc206c-66cc-4b14-df6f-a7c91adf0582"},"outputs":[{"name":"stdin","output_type":"stream","text":["Enter the dividend:  56\n","Enter the divisor:  0\n"]},{"name":"stdout","output_type":"stream","text":["Error: Arithmetic error occurred\n"]}],"source":["def perform_division(dividend, divisor):\n","    try:\n","        result = dividend / divisor\n","        print(\"Result of division:\", result)\n","    except ArithmeticError:\n","        print(\"Error: Arithmetic error occurred\")\n","\n","# Example usage:\n","try:\n","    dividend = float(input(\"Enter the dividend: \"))\n","    divisor = float(input(\"Enter the divisor: \"))\n","\n","    perform_division(dividend, divisor)\n","except ValueError:\n","    print(\"Error: Please enter valid numerical values for dividend and divisor\")\n"]},{"cell_type":"markdown","id":"8a2a2522-d940-419a-9f84-d445dc07cfa5","metadata":{"id":"8a2a2522-d940-419a-9f84-d445dc07cfa5"},"source":["9. Write a Python program that opens a file and handles a UnicodeDecodeError exception if there is an encoding issue."]},{"cell_type":"code","execution_count":null,"id":"bc6d30ca-6dcb-49b5-984c-668a2e93a7aa","metadata":{"id":"bc6d30ca-6dcb-49b5-984c-668a2e93a7aa","outputId":"10f512ea-1fe1-4f09-8976-4bf373e4e31d"},"outputs":[{"name":"stdin","output_type":"stream","text":["Enter the name of the file to read:  sample\n"]},{"name":"stdout","output_type":"stream","text":["Error: File not found\n"]}],"source":["\n","def read_file(file_name):\n","    try:\n","        with open(file_name, 'r', encoding='utf-8') as file:\n","            content = file.read()\n","            print(\"File content:\\n\", content)\n","    except FileNotFoundError:\n","        print(\"Error: File not found\")\n","    except UnicodeDecodeError:\n","        print(\"Error: Unable to decode file with UTF-8 encoding\")\n","\n","# Example usage:\n","file_name = input(\"Enter the name of the file to read: \")\n","\n","read_file(file_name)"]},{"cell_type":"code","execution_count":null,"id":"a5ee1ebf-81cf-4d71-864b-0dfa5aefb7d9","metadata":{"id":"a5ee1ebf-81cf-4d71-864b-0dfa5aefb7d9"},"outputs":[],"source":["10. Write a Python program that executes a list operation and handles an AttributeError exception if the attribute does not exist."]},{"cell_type":"code","execution_count":null,"id":"cb707fb7-b355-409c-b2cc-100781438284","metadata":{"id":"cb707fb7-b355-409c-b2cc-100781438284","outputId":"3429f0e2-c2e7-4169-9a7c-fcaec6aacfaf"},"outputs":[{"name":"stdout","output_type":"stream","text":["Length of the list: 5\n","Error: The attribute does not exist\n"]}],"source":["def list_operation(lst):\n","    try:\n","        # Assuming we want to perform some operation on the list,\n","        # such as getting its length\n","        length = len(lst)\n","        print(\"Length of the list:\", length)\n","    except AttributeError:\n","        print(\"Error: The given object does not support the 'len' operation\")\n","\n","# Example usage:\n","try:\n","    my_list = [1, 2, 3, 4, 5]\n","    list_operation(my_list)\n","\n","    # Trying to access a non-existent attribute to trigger AttributeError\n","    my_list.some_attribute\n","except AttributeError:\n","    print(\"Error: The attribute does not exist\")"]},{"cell_type":"code","execution_count":null,"id":"759818ce-5537-4014-95d3-a2ded8bb857b","metadata":{"id":"759818ce-5537-4014-95d3-a2ded8bb857b"},"outputs":[],"source":[]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.5"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":5}